File: app/src/main/java/me/zhanghai/android/files/provider/common/WalkFileTreeSearchable.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import java.io.IOException;
import java.io.InterruptedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.file.FileVisitResult;
import java8.nio.file.FileVisitor;
import java8.nio.file.Files;
import java8.nio.file.Path;
import java8.nio.file.attribute.BasicFileAttributes;
import java9.util.function.Consumer;
import me.zhanghai.android.files.util.MoreTextUtils;

public class WalkFileTreeSearchable {

    private WalkFileTreeSearchable() {}

    public static void search(@NonNull Path directory, @NonNull String query,
                              @NonNull Consumer<List<Path>> listener, long intervalMillis)
            throws IOException {
        List<Path> paths = new ArrayList<>();
        // Cannot use Files.find() or Files.walk() because it cannot ignore exceptions.
        Files.walkFileTree(directory, new FileVisitor<Path>() {
            private long mLastProgressMillis = System.currentTimeMillis();
            @NonNull
            @Override
            public FileVisitResult preVisitDirectory(@NonNull Path directory,
                                                     @NonNull BasicFileAttributes attributes)
                    throws InterruptedIOException {
                visit(directory);
                throwIfInterrupted();
                return FileVisitResult.CONTINUE;
            }
            @NonNull
            @Override
            public FileVisitResult visitFile(@NonNull Path file,
                                             @NonNull BasicFileAttributes attributes)
                    throws InterruptedIOException {
                visit(file);
                throwIfInterrupted();
                return FileVisitResult.CONTINUE;
            }
            @NonNull
            @Override
            public FileVisitResult visitFileFailed(@NonNull Path file,
                                                   @NonNull IOException exception)
                    throws InterruptedIOException {
                if (exception instanceof InterruptedIOException) {
                    throw (InterruptedIOException) exception;
                }
                exception.printStackTrace();
                visit(file);
                throwIfInterrupted();
                return FileVisitResult.CONTINUE;
            }
            @NonNull
            @Override
            public FileVisitResult postVisitDirectory(@NonNull Path directory,
                                                      @Nullable IOException exception)
                    throws InterruptedIOException {
                if (exception instanceof InterruptedIOException) {
                    throw (InterruptedIOException) exception;
                } else if (exception != null) {
                    exception.printStackTrace();
                }
                throwIfInterrupted();
                return FileVisitResult.CONTINUE;
            }
            private void visit(@NonNull Path path) {
                // Exclude the directory being searched.
                if (Objects.equals(path, directory)) {
                    return;
                }
                Path fileName = path.getFileName();
                if (fileName == null
                        || !MoreTextUtils.containsIgnoreCase(fileName.toString(), query)) {
                    return;
                }
                paths.add(path);
                long currentTimeMillis = System.currentTimeMillis();
                if (currentTimeMillis >= mLastProgressMillis + intervalMillis) {
                    listener.accept(paths);
                    mLastProgressMillis = currentTimeMillis;
                    paths.clear();
                }
            }
        });
        if (!paths.isEmpty()) {
            listener.accept(paths);
        }
    }

    private static void throwIfInterrupted() throws InterruptedIOException {
        if (Thread.interrupted()) {
            throw new InterruptedIOException();
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/WalkFileTreeSearchable.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

import java8.nio.file.DirectoryIteratorException
import java8.nio.file.FileVisitOption
import java8.nio.file.FileVisitResult
import java8.nio.file.FileVisitor
import java8.nio.file.Files
import java8.nio.file.LinkOption
import java8.nio.file.Path
import java8.nio.file.attribute.BasicFileAttributes
import java.io.IOException
import java.io.InterruptedIOException

object WalkFileTreeSearchable {
    @Throws(IOException::class)
    fun search(
        directory: Path,
        query: String,
        intervalMillis: Long,
        listener: (List<Path>) -> Unit
    ) {
        val paths = mutableListOf<Path>()
        // We cannot use Files.find() or Files.walk() because it cannot ignore exceptions.
        walkFileTreeForSearch(directory, object : FileVisitor<Path> {
            private var lastProgressMillis = System.currentTimeMillis()

            @Throws(InterruptedIOException::class)
            override fun preVisitDirectory(
                directory: Path,
                attributes: BasicFileAttributes
            ): FileVisitResult {
                visit(directory)
                throwIfInterrupted()
                return FileVisitResult.CONTINUE
            }

            @Throws(InterruptedIOException::class)
            override fun visitFile(file: Path, attributes: BasicFileAttributes): FileVisitResult {
                visit(file)
                throwIfInterrupted()
                return FileVisitResult.CONTINUE
            }

            @Throws(InterruptedIOException::class)
            override fun visitFileFailed(file: Path, exception: IOException): FileVisitResult {
                if (exception is InterruptedIOException) {
                    throw exception
                }
                exception.printStackTrace()
                visit(file)
                throwIfInterrupted()
                return FileVisitResult.CONTINUE
            }

            @Throws(InterruptedIOException::class)
            override fun postVisitDirectory(
                directory: Path,
                exception: IOException?
            ): FileVisitResult {
                if (exception is InterruptedIOException) {
                    throw exception
                }
                exception?.printStackTrace()
                throwIfInterrupted()
                return FileVisitResult.CONTINUE
            }

            private fun visit(path: Path) {
                // Exclude the directory being searched.
                if (path == directory) {
                    return
                }
                val fileName = path.fileName
                if (fileName != null && fileName.toString().contains(query, true)) {
                    paths.add(path)
                }
                if (paths.isNotEmpty()) {
                    val currentTimeMillis = System.currentTimeMillis()
                    if (currentTimeMillis >= lastProgressMillis + intervalMillis) {
                        listener(paths)
                        lastProgressMillis = currentTimeMillis
                        paths.clear()
                    }
                }
            }
        })
        if (paths.isNotEmpty()) {
            listener(paths)
        }
    }

    // This method traverses the first level first, before diving into child directories.
    // FileVisitResult returned from visitor may be ignored and always considered CONTINUE.
    @Throws(IOException::class)
    private fun walkFileTreeForSearch(start: Path, visitor: FileVisitor<in Path>): Path {
        val attributes = try {
            start.readAttributes(BasicFileAttributes::class.java)
        } catch (ignored: IOException) {
            try {
                start.readAttributes(BasicFileAttributes::class.java, LinkOption.NOFOLLOW_LINKS)
            } catch (e: IOException) {
                visitor.visitFileFailed(start, e)
                return start
            }
        }
        if (!attributes.isDirectory) {
            visitor.visitFile(start, attributes)
            return start
        }
        val directoryStream = try {
            start.newDirectoryStream()
        } catch (e: IOException) {
            visitor.visitFileFailed(start, e)
            return start
        }
        val directories = mutableListOf<Path>()
        directoryStream.use {
            visitor.preVisitDirectory(start, attributes)
            try {
                for (path in directoryStream) {
                    val attributes = try {
                        path.readAttributes(BasicFileAttributes::class.java)
                    } catch (ignored: IOException) {
                        try {
                            path.readAttributes(
                                BasicFileAttributes::class.java, LinkOption.NOFOLLOW_LINKS
                            )
                        } catch (e: IOException) {
                            visitor.visitFileFailed(path, e)
                            continue
                        }
                    }
                    visitor.visitFile(path, attributes)
                    if (attributes.isDirectory) {
                        directories.add(path)
                    }
                }
            } catch (e: DirectoryIteratorException) {
                visitor.postVisitDirectory(start, e.cause)
                return start
            }
        }
        for (path in directories) {
            Files.walkFileTree(
                path, setOf(FileVisitOption.FOLLOW_LINKS), Int.MAX_VALUE,
                object : FileVisitor<Path> {
                    @Throws(InterruptedIOException::class)
                    override fun preVisitDirectory(
                        directory: Path,
                        attributes: BasicFileAttributes
                    ): FileVisitResult {
                        if (directory == path) {
                            return FileVisitResult.CONTINUE
                        }
                        return visitor.preVisitDirectory(directory, attributes)
                    }

                    @Throws(InterruptedIOException::class)
                    override fun visitFile(
                        file: Path,
                        attributes: BasicFileAttributes
                    ): FileVisitResult {
                        if (file == path) {
                            return FileVisitResult.CONTINUE
                        }
                        return visitor.visitFile(file, attributes)
                    }

                    @Throws(InterruptedIOException::class)
                    override fun visitFileFailed(
                        file: Path,
                        exception: IOException
                    ): FileVisitResult {
                        if (file == path) {
                            // We are searching and ignoring errors, so just print it.
                            exception.printStackTrace()
                            return FileVisitResult.CONTINUE
                        }
                        return visitor.visitFileFailed(file, exception)
                    }

                    @Throws(InterruptedIOException::class)
                    override fun postVisitDirectory(
                        directory: Path,
                        exception: IOException?
                    ): FileVisitResult {
                        if (directory == path) {
                            // We are searching and ignoring errors, so just print it.
                            exception?.printStackTrace()
                            return FileVisitResult.CONTINUE
                        }
                        return visitor.postVisitDirectory(path, exception)
                    }
                }
            )
        }
        visitor.postVisitDirectory(start, null)
        return start
    }

    @Throws(InterruptedIOException::class)
    private fun throwIfInterrupted() {
        if (Thread.interrupted()) {
            throw InterruptedIOException()
        }
    }
}
